yggdrasilffi/src/lib.rs

```diff
@@ -56,257 +63,241 @@ impl From<serde_json::Error> for FFIError {
+    let json_string = serde_json::to_string(&response).unwrap();
+    CString::new(json_string).unwrap().into_raw()
+}
+
+/// Instantiates a new engine. Returns a pointer to the engine.
+///
 /// # Safety
 ///
-/// The caller is responsible for ensuring the pointer to the engine is a valid pointer,
-/// a passed null pointer is a no-op but an invalid pointer is undefined behavior.
-/// This function must be called correctly to deallocate the memory allocated for the engine in
-/// the engine_new function, failure to do so will result in a leak.
+/// The caller is responsible for freeing the allocated memory. This can be done by calling
+/// `free_engine` and passing in the pointer returned by this method. Failure to do so will result in a leak.
 #[no_mangle]
-pub unsafe extern "C" fn engine_free(ptr: *mut c_void) {
-    if ptr.is_null() {
+pub extern "C" fn new_engine() -> *mut c_void {
+    let engine = EngineState::default();
+    Box::into_raw(Box::new(engine)) as *mut c_void
+}
+
+/// Frees the memory allocated for the engine.
+///
+/// # Safety
+///
+/// The caller is responsible for ensuring the argument is a valid pointer.
+/// Null pointers will result in a no-op, but any invalid pointers will result in undefined behavior.
+/// These pointers should not be dropped for the lifetime of this function call.
+///
+/// This function must be called correctly in order to deallocate the memory allocated for the engine in
+/// the `new_engine` function. Failure to do so will result in a leak.
+#[no_mangle]
+pub unsafe extern "C" fn free_engine(engine_ptr: *mut c_void) {
+    if engine_ptr.is_null() {
         return;
     }
-    unsafe {
-        drop(Box::from_raw(ptr as *mut EngineState));
-    }
+    drop(Box::from_raw(engine_ptr as *mut EngineState));
 }
 
-/// Takes a JSON string representing a set of toggles and returns a JSON string representing
-/// the metrics that should be sent to the server
+/// Takes a JSON string representing a set of toggles. Returns a JSON string representing
+/// the metrics that should be sent to the server.
+///
 /// # Safety
 ///
-/// The caller is responsible for ensuring the pointer to the engine is a valid pointer
-/// and that the engine is not dropped while the pointer is still in use.
+/// The caller is responsible for ensuring all arguments are valid pointers.
+/// Null pointers will result in an error message being returned to the caller,
+/// but any invalid pointers will result in undefined behavior.
+/// These pointers should not be dropped for the lifetime of this function call.
 #[no_mangle]
-pub unsafe extern "C" fn engine_take_state(ptr: *mut c_void, json: *const c_char) -> *const c_char {
-    let response: Response<()> = take_state(ptr, json).into();
-    let json = serde_json::to_string(&response).unwrap();
-    CString::new(json).unwrap().into_raw()
-}
+pub extern "C" fn take_state(engine_ptr: *mut c_void, json_ptr: *const c_char) -> *const c_char {
```

Comment:
Oh this is about as safe as rubbing yourself with honey and slapping a bear

Let's put the unsafe back here! Absolutely okay for you to isolate the unsafe code block to a small portion though, I think that pattern is very nice!

Reply:
🍯🕺🐻
Loved the analogy 🤣 
Same as https://github.com/Unleash/yggdrasil/pull/47#discussion_r1348066338